_gsocblogs/2022/blog_ROOT_ZifengLuo.md

```diff
@@ -0,0 +1,105 @@
+   </tr>
+   <tr>
+      <td>std::array&lt;T, N></td>
+   </tr>
+   <tr>
+      <td>std::vector&lt;T> and ROOT::RVec&lt;T></td>
+   </tr>
+   <tr>
+      <td>std::pair&lt;T1, T2></td>
+   </tr>
+   <tr>
+      <td>std::tuple&lt;T1, …, Tn></td>
+   </tr>
+   <tr>
+      <td>Convert TTree containing user-defined classes</td>
+      <td>Any user-defined class with dictionary</td>
+      <td>Partially completed <a href="https://github.com/luozf14/TTreeToRNTuple/blob/main/src/ConverterUserDefClass.cxx">[Code]</a></td>
+   </tr>
+   <tr>
+      <td rowspan="2">Convert TTree containing branches of nested types</td>
+      <td>std::vector&lt;std::vector&lt;T> ></td>
+      <td rowspan="2">Expected by the end of Aug.</td>
+   </tr>
+   <tr>
+      <td>…</td>
+   </tr>
+   <tr>
+      <td>Tests and documentation</td>
+      <td></td>
+      <td >Expected in Sep.</td>
+   </tr>
+</table>
+
+Thanks to the [development](https://github.com/jblomer/iotools) that has been done by the ROOT I/O team, there are some programs that work as good references for me. I extended the ``gen_physlite`` program to be capable of converting TTree containing fixed-length arrays as well as variable-sized arrays. However, due to the limitation of data types supported by the current-stage RNTuple, the conversion tool is not able to deal with the multi-dimensional array. I/O team decided to mark this function as to-do until RNTuple supports multi-dimensional arrays.
+
+For TTree containing STL containers, there is no good reference. The existing ``gen_atlas`` program is dedicated to std::vector\<T\> with some specific types. I not only extended it to deal with std::vector\<T\> with all RNTuple-supported data types but also included all other STL containers supported by RNTuple (see table above).
+
+A dictionary consists of a C++ source file, which contains the type information needed by Cling and ROOT's I/O subsystem. RNTuple supports class that has a dictionary whose persistent members are themselves types with RNTuple I/O support. Currently, my code is able to migrate data stored in TTree containing user-defined classes with compiled dictionary embedded in library ``xx.so``, i.e., to do the conversion, the corresponding library must be existing. However, TTree has a feature that when it stores a class, the dictionary is also stored automatically. Hence, the next step is to find a way to load the dictionary from TTree in case the library does not exist.
+
+Converting TTree containing branches of nested types is the most difficult part, and it is expected to be complete by the end of August. Upon completion of this, we will integrate all parts into a single program that can be run as a command line tool. It will also be used as a library for TTree-to-RNTuple conversion, hence we will be working on designing the API for the library upon the completion of the first 4 parts.
```

Comment:
Perhaps also mention integration of the separate PoCs into a reusable library (+ command line tool) in the table above? (and how that affects to the schedule).